libs/api/ebpf_api.cpp

```diff
@@ -3324,18 +3334,16 @@ ebpf_object_load(_Inout_ struct bpf_object* object) NO_EXCEPT_TRY
     }
 
     if (Platform::_is_native_program(object->file_name)) {
-        struct bpf_program* program = bpf_object__next_program(object, nullptr);
-        if (program == nullptr) {
-            EBPF_RETURN_RESULT(EBPF_INVALID_ARGUMENT);
+        for (auto& program : object->programs) {
+            if (!program->autoload) {
+                continue;
+            }
+            return _ebpf_program_load_native(
```

Comment:
I think this is incorrect. `_ebpf_program_load_native()` can only be called once for a native module.

Reply:
It's not incorrect since it will return when the first execution happens.   But if there's no programs with autoload true, it won't be called.  hence the current logic.

Reply:
Ohh, I had missed the `return` part. 

In that case, if no program has `autoload` set to true, `_ebpf_program_load_native()` will not be called at all. That means the maps will also not be created. Is that expected?

I think we should always call `_ebpf_program_load_native()` and then close the program handles for the programs that have `autoload` set as false. (as this PR is already doing in `_initialize_ebpf_programs_native()`)

Reply:
> In that case, if no program has `autoload` set to true, `_ebpf_program_load_native()` will not be called at all. That means the maps will also not be created. Is that expected?

The maps ought to be created.  In fact, to match libbpf behavior, the maps should really be created _before_ trying to load the programs.  See https://github.com/libbpf/libbpf/blob/master/src/libbpf.c#L8494

Reply:
> I think we should always call `_ebpf_program_load_native()` and then close the program handles for the programs that have `autoload` set as false. (as this PR is already doing in `_initialize_ebpf_programs_native()`)

There are problems with the existing logic.  Specifically, libbpf APIs allow setting the program type on a per program basis in between open and load, so they could differ from the default based on section name.  Today `_ebpf_program_load_native()`  takes the program type and attach type as input parameters, incorrectly implying there's only one for the whole file.

Reply:
>> The maps ought to be created. In fact, to match libbpf behavior, the maps should really be created before trying to load the programs. See https://github.com/libbpf/libbpf/blob/master/src/libbpf.c#L8494

In native mode also, the maps are created before programs are loaded, but everything is done by ebpfcore. In JIT mode, ebpfsvc is supposed to be the trusted service, and it calls `get_map_descriptor_windows` during the verification process. With native mode, we cannot trust ebpfapi, hence I think the maps should also be created by ebpfcore.

>> There are problems with the existing logic. Specifically, libbpf APIs allow setting the program type on a per program basis in between open and load, so they could differ from the default based on section name. Today _ebpf_program_load_native() takes the program type and attach type as input parameters, incorrectly implying there's only one for the whole file.

This cannot be supported for native mode. Since the verification has been done offline, the program type cannot be changed after the native module has been generated (this is different than JIT mode, where program type can still be changed _before_ the program is verified)

Reply:
Updated logic here per your request.  Testing that setting program type on native is left for a future issue/PR.

Reply:
Updated this PR to test that setting program type on native fails.